functorch/csrc/DynamicLayer.cpp

```diff
@@ -499,6 +516,17 @@ void dynamicLayerFrontFallback(const c10::OperatorHandle& op, torch::jit::Stack*
       exclude = exclude.add(kBatchedKey);
     }
     hacky_include = hacky_include.add(kVmapModeKey);
+  } else if (layer.key() == DispatchKey::Functionalize) {
+    const auto args = torch::jit::last(stack, op.schema().arguments().size());
+    bool any_tensor_args = anyTensors(args, [&](const Tensor& tensor) { return true; });
+    bool any_functional_args = anyTensors(args, functionalAtCurrentLevel);
+    // Only enable dispatch on Functionalize if either:
+    // - there are any functional tensors at the current level
+    // - we hit a factory op
+    if (!any_functional_args && any_tensor_args) {
+      exclude = exclude.add(DispatchKey::Functionalize);
+    }
+    hacky_include = hacky_include.add(DispatchKey::Functionalize);
```

Comment:
Hmm, why do we need this?

We have a similar special case above for vmap, but I'm about to get rid of it because it doesn't actually work for composite operations.

Reply:
If none of the tensors are FunctionalTensors, then having the functional key enabled will mean we just skip directly to DynamicLayerBack, and then come back to DynamicLayerFront, which will check the next thing on the stack. This behavior seems OK to me but I'm probably missing something

Reply:
Hmm, I think we need it in order to functionalization to handle factory ops properly. I thought that vmap had the same problem but I might be wrong - how do you ensure that you dispatch to the batching rule properly for factory ops?

> If none of the tensors are FunctionalTensors, then having the functional key enabled will mean we just skip directly to DynamicLayerBack, and then come back to DynamicLayerFront, which will check the next thing on the stack

But what I think I want to happen here is that we hit a `Functionalize` kernel *before* getting to `DynamicLayerBack`, since what that'll do is call the underlying factory op, and then wrap the output tensor into a `FunctionalTensorWrapper`. That's basically why I wanted to keep the functionalize key in the include set here.